### PR TITLE
executors: multiqueue: publish autoscaling metrics for multiple queues

### PR DESCRIPTION
**Pull request by sanderginn**
- Closes https://github.com/sourcegraph/sourcegraph/issues/53092

~That'll do it.~ oh look, I was being overconfident.  
This PR adds a background routine that submits metrics for the total of `codeintel` and `batches` jobs.

#### Instances scaled up
Screenshot 2023-06-12 at 19 20 48

#### Both queues metrics
Screenshot 2023-06-12 at 19 27 15

#### Batches metrics
Screenshot 2023-06-12 at 19 27 26

#### Codeintel metrics
Screenshot 2023-06-12 at 19 27 35

## Test plan
- [x] Deploy nightly image to dogfood and confirm metrics submitting + autoscaling

### CODE REVIEW

**Comment by sanderginn** [comment]
Current dependencies on/for this PR:
* main
  * **PR #53197** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/53197" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/53197?utm_source=stack-comment).

**Comment by sanderginn** [comment]
@Piszmog nothing has changed compared to the other PR, needed a fresh one for the renamed branch

**Review by Piszmog — APPROVED** [review]

**Comment by sanderginn** [comment]
@Piszmog I requested a new review because the change is no longer minimal 😅

File: enterprise/internal/executor/util/multiqueue.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package util
+
+import (
+	"sort"
+	"strings"
+)
+
+func FormatQueueNamesForMetrics(queueName string, queueNames []string) string {
```
Comment (Piszmog) [inline enterprise/internal/executor/util/multiqueue.go:8]: Docs

File: enterprise/cmd/worker/internal/executorqueue/reporter.go [hunk 2]
```diff
@@ -19,6 +20,50 @@ func NewMetricReporter[T workerutil.Record](observationCtx *observation.Context,
...
+			reporters:  reporters,
+			allocation: metricsConfig.Allocations[queueName],
+		},
+		goroutine.WithName("executors.autoscaler-metrics"),
+		goroutine.WithDescription("emits metrics to GCP/AWS for auto-scaling"),
+		goroutine.WithInterval(5*time.Second),
+	), nil
+}
+
+func NewMultiqueueMetricReporter(queueNames []string, metricsConfig *Config, countFuncs ...func(ctx context.Context, includeProcessing bool) (int, error)) (goroutine.BackgroundRoutine, error) {
```
Comment (Piszmog) [inline enterprise/cmd/worker/internal/executorqueue/reporter.go:43]: docs

File: enterprise/cmd/worker/internal/executormultiqueue/multiqueue_metrics_reporter.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	dbstore "github.com/sourcegraph/sourcegraph/enterprise/cmd/worker/internal/batches"
+	"github.com/sourcegraph/sourcegraph/enterprise/cmd/worker/internal/executorqueue"
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/codeintel/autoindexing"
+	"github.com/sourcegraph/sourcegraph/internal/env"
+	"github.com/sourcegraph/sourcegraph/internal/goroutine"
+	"github.com/sourcegraph/sourcegraph/internal/observation"
+	dbworkerstore "github.com/sourcegraph/sourcegraph/internal/workerutil/dbworker/store"
+)
+
+type multiqueueMetricsReporterJob struct{}
```
Comment (Piszmog) [inline enterprise/cmd/worker/internal/executormultiqueue/multiqueue_metrics_reporter.go:17]: can you add a compile time implementation check. For example

```golang
var _ files.Store = &Client{}
```

File: enterprise/cmd/worker/internal/executormultiqueue/multiqueue_metrics_reporter.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+		return nil, err
+	}
+	codeIntelStore := dbworkerstore.New(observationCtx, db.Handle(), autoindexing.IndexWorkerStoreOptions)
+	batchesStore, err := dbstore.InitBatchSpecWorkspaceExecutionWorkerStore()
+	if err != nil {
+		return nil, err
+	}
+
+	multiqueueMetricsReporter, err := executorqueue.NewMultiqueueMetricReporter(
+		[]string{"codeintel", "batches"},
```
Comment (Piszmog) [inline enterprise/cmd/worker/internal/executormultiqueue/multiqueue_metrics_reporter.go:45]: We need a good place to centralize the possible queues somewhere 🤔 . I think this is fine. Maybe add a `TODO` to make a single place with the queues?

**Review by Piszmog — APPROVED** [review]
Nice work figuring this out! Just some minor nits.

By chance, did you also check if this works in AWS as well?

File: enterprise/internal/executor/util/multiqueue.go (hunk 1 above)
Comment (sanderginn) [reply enterprise/internal/executor/util/multiqueue.go:8]: image

**Comment by sanderginn** [comment]
@Piszmog I have not tested this on AWS yet, but given that all the implementation details are abstracted away from the cloud reporters, the only change that should be necessary is on the Terraform module side. I am confident enough to say that we'll see it work when we tackle that after the code freeze.  

I'll add a note to the issue (https://github.com/sourcegraph/sourcegraph/issues/52250).

**Pull request merged**